File: backend/pipeline_flow.py
"""Helpers for step order and resolving pipeline input (including legacy skipped steps)."""

from __future__ import annotations

from .pipeline import STEP_ORDER as _ARTICLE_STEP_ORDER


def input_source_for_step(
    step_name: str,
    statuses: dict[str, str],
    *,
    step_order: list[str] | None = None,
) -> tuple[str | None, str]:
    """
    What feeds ``step_name``'s input.

    Returns ``(source_step_key, kind)`` where *kind* is:
    - ``artifact`` — load ``source_step_key`` output
    - ``topic`` — use run topic / manual inputs (first step)
    - ``blocked`` — a required prior step is still pending
    """
    order = step_order or _ARTICLE_STEP_ORDER
    try:
        idx = order.index(step_name)
    except ValueError:
        return None, "blocked"

    for i in range(idx - 1, -1, -1):
        prev = order[i]
        st = statuses.get(prev, "pending")
        if st == "done":
            return prev, "artifact"
        if st == "skipped":
            continue
        return None, "blocked"

    return None, "topic"


def can_run_step(
    step_name: str,
    statuses: dict[str, str],
    *,
    has_topic: bool,
    step_order: list[str] | None = None,
) -> bool:
    _src, kind = input_source_for_step(step_name, statuses, step_order=step_order)
    if kind == "blocked":
        return False
    if kind == "topic":
        return has_topic
    return True

File: atlas-ui/src/utils/pipelineFlow.js
import { PIPELINE_STEP_KEYS as ARTICLE_KEYS } from "../constants/pipeline";
import { stepKeysForPipeline } from "../constants/pipelines";

/** @typedef {{ kind: 'artifact', stepKey: string } | { kind: 'topic' } | { kind: 'blocked' }} InputSource */

/** @param {string} stepKey @param {Record<string, string>} statuses @returns {InputSource} */
export function inputSourceForStep(stepKey, statuses, pipelineId = null) {
  const keys = pipelineId ? stepKeysForPipeline(pipelineId) : ARTICLE_KEYS;
  const idx = keys.indexOf(stepKey);
  if (idx < 0) return { kind: "blocked" };

  for (let i = idx - 1; i >= 0; i -= 1) {
    const prev = keys[i];
    const st = statuses[prev] || "pending";
    if (st === "done") return { kind: "artifact", stepKey: prev };
    if (st === "skipped") continue;
    return { kind: "blocked" };
  }
  return { kind: "topic" };
}

export function canRunStep(stepKey, statuses, topic, pipelineId = null) {
  const src = inputSourceForStep(stepKey, statuses, pipelineId);
  if (src.kind === "blocked") return false;
  if (src.kind === "topic") return Boolean(String(topic || "").trim());
  return true;
}
